source/blog/2018-01-22-CICD-workflow-with-zuul.html.md

```diff
@@ -0,0 +1,407 @@
+            dependencies:
+              - sf-ci-functional-minimal
+              - sf-ci-upgrade-minimal
+              - sf-ci-functional-allinone
+              - sf-ci-upgrade-allinone
+```
+
+The deployment pipeline needs to use the *queue* option to group all the
+approved changes in dependent order.
+When multiple changes are approved in parallel, they will all be tested
+together before being merged, as if they were submitted with a
+depends-on relationship.
+
+The deployment pipeline is similar to the integration pipeline, it just adds
+a publish job that will only run if all the integration tests succeed.
+This ensures that changes are consistently tested with the projects' current
+state before being deployed.
+
+
+## Deployment job definition
+
+The job is declared in a zuul.yaml file as below:
+
+```yaml
+- job:
+    name: sf-rpm-publish
+    description: Publish Software Factory rpm to koji
+    run: playbooks/rpmpublish.yaml
+    hold-following-changes: true
+    required-projects:
+      - software-factory/sfinfo
+    secrets:
+      - sf_koji_configuration
+```
+
+This job is using the *hold-following-changes* setting to ensure that only
+the top of the gate gets published. If the deployement is happening in the
+post or release pipeline, then this setting can be replaced by a *semaphore*
+instead, for example:
+
+```yaml
+- job:
+    name: deployment
+    semaphore: production-access
+
+- semaphore:
+    name: production-access
+    max: 1
+```
+
+This prevents concurrency issues when multiple changes are approved in parallel.
+
+
+# Zuul concepts summary
+
+This article covered the following concepts:
+
+* Project types:
+
+  * *config-projects*: hold deployment secrets and set projects' pipelines.
+  * *untrusted-projects*: the projects being tested and deployed.
+
+* Playbook variables:
+
+  * *zuul.projects*: the projects installed on the test instance,
+  * *zuul.items*: the list of changes being tested with depends-on,
+  * *zuul.executor.log_root*: the location of job artifacts, and
+  * *zuul_return*: an Ansible module to share data between jobs.
+
+* Job options:
+
+  * *required-projects*: the list of projects to copy on the test instance,
+  * *dependencies*: the list of jobs to wait for,
+  * *secret*: the deployment job's secret,
+  * *post-review*: prevents a job from running speculatively,
+  * *hold-following-changes*: makes dependent pipelines run in serial, and
+  * *semaphore*: prevents concurrent deployment of different changes.
+
+* Pipeline options:
+
+  * Job settings can be modified per project, and
+  * *queue* makes all the projects depend on each other automatically.
+
+# Conclusion
+
+Zuul can be used to effectively manage complex continous integration and
+deployment pipelines with powerfull cross repository dependencies management.
+
+This article presented the Software Factory workflow where rpm packages are
+being continously built, tested and delivered through code review.
```

Comment:
Line 404, 405 could also be put at the beginning of the article.

Reply:
Done, thanks for the review Fabien!